File: stylus.go
package main

import (
	"unsafe"

	. "github.com/linux-surface/iptsd/protocol"
)

func IptsStylusHandleData(ipts *IptsContext, data IptsStylusData) error {
	stylus := ipts.Devices.ActiveStylus

	prox := (data.Mode & IPTS_STYLUS_REPORT_MODE_PROX) >> 0
	touch := (data.Mode & IPTS_STYLUS_REPORT_MODE_TOUCH) >> 1
	button := (data.Mode & IPTS_STYLUS_REPORT_MODE_BUTTON) >> 2
	rubber := (data.Mode & IPTS_STYLUS_REPORT_MODE_RUBBER) >> 3

	btn_pen := prox * (1 - rubber)
	btn_rubber := prox * rubber

	sx, sy := stylus.Processor.Smooth(int(data.X), int(data.Y))
	tx, ty := stylus.Processor.Tilt(int(data.Altitude), int(data.Azimuth))

	if prox == 0 {
		stylus.Processor.Flush()
	}

	stylus.Device.Emit(EV_KEY, BTN_TOUCH, int32(touch))
	stylus.Device.Emit(EV_KEY, BTN_TOOL_PEN, int32(btn_pen))
	stylus.Device.Emit(EV_KEY, BTN_TOOL_RUBBER, int32(btn_rubber))
	stylus.Device.Emit(EV_KEY, BTN_STYLUS, int32(button))

	stylus.Device.Emit(EV_ABS, ABS_X, int32(sx))
	stylus.Device.Emit(EV_ABS, ABS_Y, int32(sy))
	stylus.Device.Emit(EV_ABS, ABS_PRESSURE, int32(data.Pressure))
	stylus.Device.Emit(EV_ABS, ABS_MISC, int32(data.Timestamp))

	stylus.Device.Emit(EV_ABS, ABS_TILT_X, int32(tx))
	stylus.Device.Emit(EV_ABS, ABS_TILT_Y, int32(ty))

	err := stylus.Device.Emit(EV_SYN, SYN_REPORT, 0)
	if err != nil {
		return err
	}

	return nil
}

func IptsStylusHandleSerialChange(ipts *IptsContext, serial uint32) error {
	for _, stylus := range ipts.Devices.Styli {
		if stylus.Serial != serial {
			continue
		}

		ipts.Devices.ActiveStylus = stylus
		return nil
	}

	/*
	 * Before touching the screen for the first time, the stylus
	 * will report its serial as 0. Once you touch the screen,
	 * the serial will be reported correctly until you restart
	 * the machine.
	 */
	if ipts.Devices.ActiveStylus.Serial == 0 {
		ipts.Devices.ActiveStylus.Serial = serial
		return nil
	}

	err := ipts.Devices.AddStylus(ipts, serial)
	if err != nil {
		return err
	}

	return nil
}

func IptsStylusHandleReportSerial(ipts *IptsContext) error {
	report, err := ipts.Protocol.ReadStylusReportSerial()
	if err != nil {
		return err
	}

	if ipts.Devices.ActiveStylus.Serial != report.Serial {
		err = IptsStylusHandleSerialChange(ipts, report.Serial)
		if err != nil {
			return err
		}
	}

	for i := uint8(0); i < report.Elements; i++ {
		data, err := ipts.Protocol.ReadStylusData()
		if err != nil {
			return err
		}

		err = IptsStylusHandleData(ipts, data)
		if err != nil {
			return err
		}
	}

	return nil
}

func IptsStylusHandleReportTilt(ipts *IptsContext) error {
	report, err := ipts.Protocol.ReadStylusReport()
	if err != nil {
		return err
	}

	for i := uint8(0); i < report.Elements; i++ {
		data, err := ipts.Protocol.ReadStylusData()
		if err != nil {
			return err
		}

		err = IptsStylusHandleData(ipts, data)
		if err != nil {
			return err
		}
	}

	return nil
}

func IptsStylusHandleReportNoTilt(ipts *IptsContext) error {
	report, err := ipts.Protocol.ReadStylusReportSerial()
	if err != nil {
		return err
	}

	if ipts.Devices.ActiveStylus.Serial != report.Serial {
		err = IptsStylusHandleSerialChange(ipts, report.Serial)
		if err != nil {
			return err
		}
	}

	for i := uint8(0); i < report.Elements; i++ {
		data, err := ipts.Protocol.ReadStylusDataNoTilt()
		if err != nil {
			return err
		}

		fullData := IptsStylusData{
			Mode:      uint16(data.Mode),
			X:         data.X,
			Y:         data.Y,
			Pressure:  data.Pressure * 4,
			Altitude:  0,
			Azimuth:   0,
			Timestamp: 0,
		}

		err = IptsStylusHandleData(ipts, fullData)
		if err != nil {
			return err
		}
	}

	return nil
}

func IptsStylusHandleInput(ipts *IptsContext, frame IptsPayloadFrame) error {
	size := uint32(0)

	for size < frame.Size {
		report, err := ipts.Protocol.ReadReport()
		if err != nil {
			return err
		}

		size += uint32(report.Size) + uint32(unsafe.Sizeof(report))

		switch report.Type {
		case IPTS_REPORT_TYPE_STYLUS_NO_TILT:
			err = IptsStylusHandleReportNoTilt(ipts)
			break
		case IPTS_REPORT_TYPE_STYLUS_TILT:
			err = IptsStylusHandleReportTilt(ipts)
			break
		case IPTS_REPORT_TYPE_STYLUS_TILT_SERIAL:
			err = IptsStylusHandleReportSerial(ipts)
			break
		default:
			// ignored
			err = ipts.Protocol.Skip(uint32(report.Size))
			break
		}

		if err != nil {
			return err
		}
	}

	return nil
}

File: stylus.c
// SPDX-License-Identifier: GPL-2.0-or-later

#include <linux/uinput.h>
#include <stdint.h>
#include <stdio.h>

#include "context.h"
#include "devices.h"
#include "protocol.h"
#include "stylus-processing.h"
#include "utils.h"

static int iptsd_stylus_handle_data(struct iptsd_context *iptsd,
		struct ipts_stylus_data data)
{
	int sx, sy;
	int tx, ty;

	struct iptsd_stylus_device *stylus = iptsd->devices.active_stylus;

	int prox = (data.mode & IPTS_STYLUS_REPORT_MODE_PROX) >> 0;
	int touch = (data.mode & IPTS_STYLUS_REPORT_MODE_TOUCH) >> 1;
	int button = (data.mode & IPTS_STYLUS_REPORT_MODE_BUTTON) >> 2;
	int rubber = (data.mode & IPTS_STYLUS_REPORT_MODE_RUBBER) >> 3;

	int btn_pen = prox * (1 - rubber);
	int btn_rubber = prox * rubber;

	iptsd_stylus_processing_smooth(&stylus->processor,
			data.x, data.y, &sx, &sy);

	iptsd_stylus_processing_tilt(&stylus->processor,
			data.altitude, data.azimuth, &tx, &ty);

	if (prox == 0)
		iptsd_stylus_processing_flush(&stylus->processor);

	iptsd_devices_emit(stylus->dev, EV_KEY, BTN_TOUCH, touch);
	iptsd_devices_emit(stylus->dev, EV_KEY, BTN_TOOL_PEN, btn_pen);
	iptsd_devices_emit(stylus->dev, EV_KEY, BTN_TOOL_RUBBER, btn_rubber);
	iptsd_devices_emit(stylus->dev, EV_KEY, BTN_STYLUS, button);

	iptsd_devices_emit(stylus->dev, EV_ABS, ABS_X, sx);
	iptsd_devices_emit(stylus->dev, EV_ABS, ABS_Y, sy);
	iptsd_devices_emit(stylus->dev, EV_ABS, ABS_PRESSURE, data.pressure);
	iptsd_devices_emit(stylus->dev, EV_ABS, ABS_MISC, data.timestamp);

	iptsd_devices_emit(stylus->dev, EV_ABS, ABS_TILT_X, tx);
	iptsd_devices_emit(stylus->dev, EV_ABS, ABS_TILT_Y, ty);

	int ret = iptsd_devices_emit(stylus->dev, EV_SYN, SYN_REPORT, 0);
	if (ret < 0)
		iptsd_err(ret, "Failed to emit stylus report");

	return ret;
}

static int iptsd_stylus_change_serial(struct iptsd_context *iptsd,
		uint32_t serial)
{
	for (int i = 0; i < IPTSD_MAX_STYLI; i++) {
		if (iptsd->devices.styli[i].serial != serial)
			continue;

		iptsd->devices.active_stylus = &iptsd->devices.styli[i];
		return 0;
	}

	/*
	 * Before touching the screen for the first time, the stylus
	 * will report its serial as 0. Once you touch the screen,
	 * the serial will be reported correctly until you restart
	 * the machine.
	 */
	if (iptsd->devices.active_stylus->serial == 0) {
		iptsd->devices.active_stylus->serial = serial;
		return 0;
	}

	int ret = iptsd_devices_add_stylus(&iptsd->devices, serial);
	if (ret < 0)
		iptsd_err(ret, "Failed to add new stylus");

	return ret;
}

static int iptsd_stylus_handle_tilt_serial(struct iptsd_context *iptsd,
		struct ipts_report *report)
{
	int pos = 0;
	struct ipts_stylus_report_serial *sreport =
		(struct ipts_stylus_report_serial *)report->data;

	if (iptsd->devices.active_stylus->serial != sreport->serial) {
		int ret = iptsd_stylus_change_serial(iptsd, sreport->serial);
		if (ret < 0) {
			iptsd_err(ret, "Failed to change stylus");
			return ret;
		}
	}

	for (int i = 0; i < sreport->elements; i++) {
		struct ipts_stylus_data *data =
			(struct ipts_stylus_data *)&sreport->data[pos];

		int ret = iptsd_stylus_handle_data(iptsd, *data);
		if (ret < 0) {
			iptsd_err(ret, "Failed to handle stylus report");
			return ret;
		}
	}

	return 0;
}

static int iptsd_stylus_handle_tilt(struct iptsd_context *iptsd,
		struct ipts_report *report)
{
	int pos = 0;
	struct ipts_stylus_report *sreport =
		(struct ipts_stylus_report *)report->data;

	for (int i = 0; i < sreport->elements; i++) {
		struct ipts_stylus_data *data =
			(struct ipts_stylus_data *)&sreport->data[pos];

		int ret = iptsd_stylus_handle_data(iptsd, *data);
		if (ret < 0) {
			iptsd_err(ret, "Failed to handle stylus report");
			return ret;
		}
	}

	return 0;
}

static int iptsd_stylus_handle_no_tilt(struct iptsd_context *iptsd,
		struct ipts_report *report)
{
	int pos = 0;
	struct ipts_stylus_data full_data;

	struct ipts_stylus_report_serial *sreport =
		(struct ipts_stylus_report_serial *)report->data;

	if (iptsd->devices.active_stylus->serial != sreport->serial) {
		int ret = iptsd_stylus_change_serial(iptsd, sreport->serial);
		if (ret < 0) {
			iptsd_err(ret, "Failed to change stylus");
			return ret;
		}
	}

	for (int i = 0; i < sreport->elements; i++) {
		struct ipts_stylus_data_no_tilt *data =
			(struct ipts_stylus_data_no_tilt *)&sreport->data[pos];

		full_data.mode = data->mode;
		full_data.x = data->x;
		full_data.y = data->y;
		full_data.pressure = data->pressure * 4;
		full_data.altitude = 0;
		full_data.azimuth = 0;
		full_data.timestamp = 0;

		int ret = iptsd_stylus_handle_data(iptsd, full_data);
		if (ret < 0) {
			iptsd_err(ret, "Failed to handle stylus report");
			return ret;
		}
	}

	return 0;
}

int iptsd_stylus_handle_input(struct iptsd_context *iptsd,
		struct ipts_payload_frame *frame)
{
	int pos = 0;

	while (pos < frame->size) {
		int ret = 0;
		struct ipts_report *report =
			(struct ipts_report *)&frame->data[pos];

		switch (report->type) {
		case IPTS_REPORT_TYPE_STYLUS_NO_TILT:
			ret = iptsd_stylus_handle_no_tilt(iptsd, report);
			break;
		case IPTS_REPORT_TYPE_STYLUS_TILT:
			ret = iptsd_stylus_handle_tilt(iptsd, report);
			break;
		case IPTS_REPORT_TYPE_STYLUS_TILT_SERIAL:
			ret = iptsd_stylus_handle_tilt_serial(iptsd, report);
			break;
		}

		if (ret < 0) {
			iptsd_err(ret, "Failed to handle stylus input");
			return ret;
		}

		pos += report->size + sizeof(struct ipts_report);
	}

	return 0;
}
